### PR TITLE
fix InitVar problem

### PR DESCRIPTION
**Pull request by Czaki**
# Description

Fixes type detection requested in #1940.

This PR adds to changes from #1966 support for InitVar and Optional. Checks for optional are a little tricky. 

## Type of change

- [x] Bug-fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)

# References

# How has this been tested?

- [ ] example: the test suite for my feature covers cases x, y, and z
- [x] example: all tests pass with my change

## Final checklist:
- [x] My PR is the minimum possible work for the desired functionality
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

**Comment by Czaki** [comment]
@tlambert03 please check. I add conservative support for `Optional` Which depends on the current implementation of its, but its change should not cause the crash of napari application. But it is not pretty.

File: napari/utils/events/dataclass.py [hunk 1]
```diff
@@ -230,6 +230,18 @@ def _type_to_compare(type_) -> Optional[Callable[[Any, Any], bool]]:
...
+    if (
+        hasattr(type_, "__origin__")
+        and hasattr(type_.__origin__, "_name")
+        and type_.__origin__._name == "Union"
+        and hasattr(type_, "__args__")
+        and isinstance(type_.__args__, tuple)
+        and len(type_.__args__) == 2
+        and isinstance(None, type_.__args__[1])
+    ):
+        type_ = type_.__args__[0]
```
Comment (tlambert03) [inline napari/utils/events/dataclass.py]:
```suggestion
    _args = get_args(type_)
    if get_origin(type_) is Union and len(_args) == 2 and isinstance(None, _args[1]):
        type_ = _args[0]
```

(then add `from typing import Union`)

Comment (tlambert03) [reply napari/utils/events/dataclass.py]: can you also add a short docstring to `_type_to_compare`?

Comment (tlambert03) [reply napari/utils/events/dataclass.py]: oh, and add a comment above that line, saying "get main type from Optional[type]" so it's clear what's going on there

**Review by tlambert03 — APPROVED** [review]

File: napari/utils/events/dataclass.py (hunk 1 above)
Comment (Czaki) [reply napari/utils/events/dataclass.py]: Could you check now?

**Review by tlambert03 — APPROVED** [review]
great! thanks!

File: napari/utils/events/dataclass.py (hunk 1 above)
Comment (Czaki) [reply napari/utils/events/dataclass.py]: I do not know `get_origin` and `get_args`. Thanks for pointing it.

**Review by sofroniewn — APPROVED** [review]
Thanks for the fix with a test @Czaki and review @tlambert03 - this looks good to me. I will merge now

**Comment by sofroniewn** [comment]
cc @kevinyamauchi this has now merge and should unblock #1940

**Pull request merged**